unit_tests/src/test_atomic_apply_acl_operations.cpp

```diff
@@ -636,6 +636,109 @@ TEST_CASE("#7408: Atomic acls respect entity zone")
+                                         std::make_pair("read_object", fs::perms::read),
+                                         std::make_pair("create_metadata", fs::perms::create_metadata),
+                                         std::make_pair("modify_metadata", fs::perms::modify_metadata),
+                                         std::make_pair("delete_metadata", fs::perms::delete_metadata),
+                                         std::make_pair("create_object", fs::perms::create_object),
+                                         std::make_pair("modify_object", fs::perms::write),
+                                         std::make_pair("delete_object", fs::perms::delete_object) };
+
+    std::vector<char*> err_strs;
+    err_strs.reserve(new_acls_and_perms.size());
+    irods::at_scope_exit free_errstrings{[&err_strs] {
+        for (auto&& e : err_strs) {
+            // NOLINTNEXTLINE(cppcoreguidelines-no-malloc,cppcoreguidelines-owning-memory)
+            std::free(e);
+        }
+    }};
+
+    for(const auto& acl_and_perm : new_acls_and_perms) {
+        const auto json_input = json{
+            {"logical_path", user_home},
+            {"operations", json::array({
+                {
+                    {"entity_name", acl_user_name},
+                    {"acl", acl_and_perm.first}
+                }
+            })}
+        }.dump();
+
+        // clang-format on
+
+        // clean up acl after each iteration
+        irods::at_scope_exit remove_acls{[&] {
+            // clang-format off
+        const auto json_input = json{
+            {"logical_path", user_home},
+            {"operations", json::array({
+                {
+                    {"entity_name", acl_user.name},
+                    {"acl", "null"}
+                },
+            })}
+        }.dump();
+            // clang-format on
+
+            char* json_error_string{};
+
+            // NOLINTNEXTLINE(cppcoreguidelines-no-malloc,cppcoreguidelines-owning-memory)
+            irods::at_scope_exit free_memory{[&json_error_string] { std::free(json_error_string); }};
+
+            REQUIRE(rc_atomic_apply_acl_operations(
+                        static_cast<rcComm_t*>(conn), json_input.c_str(), &json_error_string) == 0);
+            REQUIRE(json_error_string == "{}"s);
+
+            const auto perms = fs::client::status(conn, user_home).permissions();
+            auto b = std::begin(perms);
+            auto e = std::end(perms);
+            REQUIRE(std::find_if(
+                        b, e, [&acl_user](const fs::entity_permission& p) { return p.name == acl_user.name; }) == e);
+        }};
+
+        char* json_error_string{};
+
+        REQUIRE(rc_atomic_apply_acl_operations(static_cast<rcComm_t*>(conn), json_input.c_str(), &json_error_string) ==
+                0);
+        REQUIRE(json_error_string == "{}"s);
+        err_strs.push_back(json_error_string);
```

Comment:
Do we need to put this before the `REQUIRE(json_error_string == "{}"s);`? In the event that the assertion fails, the string will leak.

Reply:
Actually, will the string be populated in the event that `rc_atomic_apply_acl_operations` returns non-zero? We might need to use an `at_scope_exit` in the loop to ensure the string is freed.

Reply:
I could initialize it to 0, add the `char*` to the vector beforehand and then check that it's nonzero before freeing in the outer `at_scope_exit`. Do you think that would work?

Reply:
Yes, I think that works.